nova/scheduler/weights/vm_hv_ratio.py

```diff
@@ -0,0 +1,58 @@
+# Copyright (c) 2021 SAP SE
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+"""
+VmHvRatioWeigher. Weigh hosts by how well they fit the VM size.
+"""
+from oslo_log import log as logging
+
+import nova.conf
+from nova.scheduler.mixins import HypervisorSizeMixin
+from nova.scheduler import weights
+
+LOG = logging.getLogger(__name__)
+
+CONF = nova.conf.CONF
+
+
+class VmHvRatioWeigher(weights.BaseHostWeigher, HypervisorSizeMixin):
+    minval = 0
+
+    def __init__(self, *args, **kwargs):
+        weights.BaseHostWeigher.__init__(self, *args, **kwargs)
+        HypervisorSizeMixin.__init__(self)
+
+    def weight_multiplier(self):
+        """Override the weight multiplier."""
+        return CONF.filter_scheduler.vm_hv_ratio_weight_multiplier
+
+    def _weigh_object(self, host_state, request_spec):
+        """We want best fit
+
+        Best fit is smallest number of VMs that can fit the HV size.
+
+        If we get the number of VMs that can fit the HV by HV / VM, we get a
+        higher number for less well fitting. Since we want a higher number for
+        better fitting, we use the reciprocal value (VM / HV), which is also
+        the ratio of the VM size to the HV size.
+        """
+        requested_ram_mb = request_spec.memory_mb
+
+        hypervisor_ram_mb = self._get_hv_size(host_state)
+        if not hypervisor_ram_mb:
+            LOG.debug('Cannot retrieve hypervisor size for %(host_state)s.',
+                      {'host_state': host_state})
+            return 0.0
+
+        return requested_ram_mb / hypervisor_ram_mb
```

Comment:
I would suggest also to return 0.0, when `requested_ram_mb / hypervisor_ram_mb > 1` . It should never happen, as the placement-api should take care of that, but just in case.

Reply:
OK. I added debug log for that, too.